core/src/main/java/com/yahoo/imapnio/async/data/ExtendedSearchSequence.java

```diff
@@ -0,0 +1,192 @@
+package com.yahoo.imapnio.async.data;
+
+import com.sun.mail.iap.Argument;
+import com.sun.mail.imap.protocol.SearchSequence;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.mail.search.AndTerm;
+import javax.mail.search.FlagTerm;
+import javax.mail.search.NotTerm;
+import javax.mail.search.OrTerm;
+import javax.mail.search.SearchException;
+import javax.mail.search.SearchTerm;
+
+import java.io.IOException;
+
+/**
+ * This class extends the search sequence for modification sequence with the optional fields entry
+ * name and type defined in https://tools.ietf.org/html/rfc7162#section-3.1.5.
+ */
+public class ExtendedSearchSequence extends SearchSequence {
+
+    /**
+     * Generate the IMAP search sequence for the given search expression.
+     *
+     * @param term the search term
+     * @return the IMAP search sequence argument
+     * @throws SearchException will not throw
+     * @throws IOException will not throw
+     */
+    public Argument generateSequence(@Nonnull final SearchTerm term) throws IOException, SearchException {
+        return generateSequence(term, null);
+    }
+
+    /**
+     * Generate the IMAP search sequence for the given search expression.
+     *
+     * @param term the search term
+     * @param charset the character set
+     * @return the IMAP search sequence argument
+     * @throws SearchException will not throw
+     * @throws IOException will not throw
+     */
+    @Override
+    public Argument generateSequence(@Nonnull final SearchTerm term, @Nullable final String charset)
+            throws SearchException, IOException {
+        if (term instanceof ExtendedModifiedSinceTerm) {
+            return modifiedSince((ExtendedModifiedSinceTerm) term);
+        } else if (term instanceof AndTerm) {
+            return and((AndTerm) term, charset);
```

Comment:
Why are you not calling super.generateSequence(..) ?

Reply:
super.generateSequence will always use ModifiedSinceTerm instead of ExtendedModifiedSinceTerm when using And, Or, Not operation + MODSEQ.

For example, SEARCH OR NOT MODSEQ 720162338 LARGER 50000

Reply:
Since this library will always use ExtendedSearchSequence and not SearchSequence, calling super.generateSequence() for terms which is not ExtendedModifiedSinceTerm, the recursive call to generateSequence() from any other term should end up in this function and not the base class function.

Reply:
Only ExtendedModifiedSinceTerm must use ExtendedSearchSequence. And, Or, and Not may or may not use ExtendedSearchSequence; however, we cannot know whether it will have ExtendedModifiedSinceTerm or not unless we reach the leaf node of the search tree. Therefore, only for And, Or, and Not, we have to use ExtendedSearchSequence to search to avoid the mismatch.